### PR TITLE
[LTR] Missing param handling improvements

### PR DESCRIPTION
**Pull request by afoucret**
Following up on: https://github.com/elastic/elasticsearch/pull/102601

This PR implements add the following to the learn to rank config stored within the model:
- default param values at the global level 
- default score for each query extractor

The following config: 

```
"learn_to_rank": {
  "default_params": {
    "category_id": 12
  },
  "feature_extractors": [
    { "query_extractor": { "default_score": 2, "feature_name": "product_match", "query": {  "match": { "title" : "{{fulltext_search}}" } } } },
    { "query_extractor": { "feature_name": "category_match", "query": {  "term": { "category" : "{{category_id}}" } } } },
    { "query_extractor": { "feature_name": "brand_match", "query": {  "match": { "brand" : "{{fulltext_search}}" } } } }
  ]
}
```

```
  "feature_extractors": [
     { "query_extractor": { "feature_name": "product_match", "query": {  "match_all": { "boost": 2 } } } } },
     { "query_extractor": { "feature_name": "category_match", "query": {  "term": { "category" : "12" } } } }
     { "query_extractor": { "feature_name": "brand_match", "query": {  "match_none: {} } } } }
  ]
```

Explanation:
- `category_match` feature will be using the default param for the missing {{category_id}} param
- `product_match` feature will be using a query that is returning a score of 2 (default_score) for each document (boosted match_all query) as a replacement the query with the {{fulltext_search}} missing param
- `brand_match` will be using a match_none query (feature extracted equal to 0) as a replacement as a replacement the query with the {{fulltext_search}} missing param

### CODE REVIEW

**Review by benwtrent — COMMENTED** [review]
I like it. Some minor comments.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/ltr/QueryExtractorBuilder.java [hunk 1]
```diff
@@ -55,18 +64,24 @@ public static QueryExtractorBuilder fromXContent(XContentParser parser, Object c
...
     public QueryExtractorBuilder(StreamInput input) throws IOException {
-        this(input.readString(), QueryProvider.fromStream(input));
+        this(input.readString(), QueryProvider.fromStream(input), input.readFloat());
     }
 
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject();
         builder.field(FEATURE_NAME.getPreferredName(), featureName);
+        builder.field(DEFAULT_SCORE.getPreferredName(), defaultScore);
```
[the remark this answers is not in the document]
Comment (afoucret) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/ltr/QueryExtractorBuilder.java]: It makes sense to me. I did add a conditional.

**Review by kderusso — COMMENTED** [review]
Overall changes LGTM, might need some more BWC work to get tests to pass

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/ltr/QueryExtractorBuilderTests.java [hunk 2]
```diff
@@ -26,7 +26,7 @@ public class QueryExtractorBuilderTests extends AbstractXContentSerializingTestC
     protected boolean lenient;
 
     public static QueryExtractorBuilder randomInstance() {
-        return new QueryExtractorBuilder(randomAlphaOfLength(10), QueryProviderTests.createRandomValidQueryProvider());
+        return new QueryExtractorBuilder(randomAlphaOfLength(10), QueryProviderTests.createRandomValidQueryProvider(), randomFloat());
```
[the remark this answers is not in the document]
Comment (afoucret) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/ltr/QueryExtractorBuilderTests.java]: Fixed!

**Review by benwtrent — APPROVED** [review]

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/LearnToRankConfig.java [hunk 3]
```diff
@@ -33,7 +36,9 @@ public class LearnToRankConfig extends RegressionConfig implements Rewriteable<L
     static final TransportVersion MIN_SUPPORTED_TRANSPORT_VERSION = TransportVersion.current();
     public static final ParseField NUM_TOP_FEATURE_IMPORTANCE_VALUES = new ParseField("num_top_feature_importance_values");
     public static final ParseField FEATURE_EXTRACTORS = new ParseField("feature_extractors");
-    public static LearnToRankConfig EMPTY_PARAMS = new LearnToRankConfig(null, null);
+    public static final ParseField DEFAULT_PARAMS = new ParseField("default_params");
```
[the remark this answers is not in the document]
Comment (afoucret) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/LearnToRankConfig.java:39]: I did update the PR description to reflect the change.

**Pull request merged**